Support fixes for destructuring, default exports, and heritage clauses

h-joo opened the pull request:
Code review will be easier to be done commit-by-commit.

This PR intends to support fixes for the following in the language-service fixer for type annotation

1. object, array destructuring
2. default exports
3. heritage clauses

Upcoming work:

1. Properly support the type generation for`Symbol()`s
2. Wrap the code-mod around the language service instead. I tried a bit but there were some files (external-declarations/tests/source/binder/re-exported-import-visibility.ts) that was causing module imports thus the language service itself failed, so I had to revert my changes.
3. Improve Diagnostics handling. Currently it occasionally prints multiple diagnostics for one thing. (e.g. For `export const c = [...foo()];` , it has diagnostics for `c` and for `foo`, thus the fixer also generates multiple fixes for the same thing.

h-joo committed:
Support default exports

in the fixMissingTypeAnnotation language service.

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
[Non-functional] Restructure the fixer

to be able to create change outside the node itself

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Support fixes for mixin heritage clauses

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Support fixes for destructuring patterns

we split destructuring patterns as the declaration list
is splitted on d.ts emit

Signed-off-by: Hana Joo <hanajoo@google.com>

h-joo committed:
Add a test case for Symbol

Signed-off-by: Hana Joo <hanajoo@google.com>

dragomirtitian commented on tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports12.ts:
```diff
@@ -0,0 +1,30 @@
+`function mixin<T extends new (...a: any) => any>(ctor: T): T {
+    return ctor;
+}
+class Point2D { x = 0; y = 0; }
```
Why can't we use the same approach here as for class declarations? 

```ts
class Point2D { x = 0; y = 0; }
const Point3DBase: typeof Point2D = (mixin(Point2D));
export const Point3D = class extends Point3DBase {  z = 0; }`
});
```

h-joo replied:
It's because the diagnostics occur on `Point3D`, not on the `extends` clause. 
I can try to make this pattern recognized by the fixer, but potentially having a diagnostic on Point3D and a nested information on extends. Either way, I'll try this out in following PRs.

dragomirtitian commented on tests/cases/fourslash/codeFixMissingTypeAnnotationOnExports16.ts:
```diff
@@ -0,0 +1,26 @@
+}
+const dest = foo();
+export const x: 1 = dest.x;
+const temp: 1 | 0 = dest.y;
```
Is the annotation here required ?

h-joo replied:
No, deleted.

dragomirtitian commented on src/services/codefixes/fixMissingTypeAnnotationOnExports.ts:
```diff
@@ -159,23 +188,248 @@ function addTypeAnnotationOnNode(node: Node, typeChecker: TypeChecker): Node | N
+    IDENTIFIER = 3,
+}
+
+type SubExpression = {kind: ExpressionType.TEXT, text: string} |
+                     {kind: ExpressionType.COMPUTED, computed: Expression} |
+                     {kind: ExpressionType.ARRAY_ACCESS, arrayIndex: number} |
+                     {kind: ExpressionType.IDENTIFIER, identifier: Identifier} ;
```
Formatting of unions would usually put the `|` in front:
Ex:

```ts
export type LateVisibilityPaintedStatement =
    | AnyImportSyntax
    | VariableStatement
    | ClassDeclaration
    | FunctionDeclaration
    | ModuleDeclaration
    | TypeAliasDeclaration
    | InterfaceDeclaration
    | EnumDeclaration;
    
```

h-joo replied:
Done

h-joo committed:
Fix PR Comments

1. Reformat union types
2. Do not add type annotation when not needed
3. Fix tests for #2 changes.

Signed-off-by: Hana Joo <hanajoo@google.com>
